### PR TITLE
SONARJAVA-1465 Fail fast in case of too many program states

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java [hunk 1]
```diff
@@ -73,7 +113,7 @@ public boolean equals(Object obj) {
 
   private boolean equalsRelation(BinaryRelation rel) {
     if (kind.equals(rel.kind)) {
-      return key[0] == rel.key[0] && key[1] == rel.key[1];
+      return leftOp == rel.leftOp && rightOp == rel.rightOp;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)
MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java [hunk 2]
```diff
@@ -35,6 +35,49 @@
 
 public abstract class BinaryRelation {
 
+
+  public static BinaryRelation binaryRelation(Kind kind, SymbolicValue leftOp, SymbolicValue rightOp) {
+    BinaryRelation relation;
+    switch (kind) {
+      case EQUAL:
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java]: MAJOR Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)

File: java-squid/src/main/java/org/sonar/java/se/symbolicvalues/NotEqualRelation.java [hunk 3]
```diff
@@ -31,12 +31,18 @@
...
+      symmetric = binaryRelation(RelationalSymbolicValue.Kind.NOT_EQUAL, rightOp, leftOp);
+    }
+    return symmetric;
   }
 
   @Override
   public BinaryRelation inverse() {
-    return new EqualRelation(leftOp, rightOp);
+    if(inverse == null) {
+      inverse = BinaryRelation.binaryRelation(RelationalSymbolicValue.Kind.EQUAL, leftOp, rightOp);;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/symbolicvalues/NotEqualRelation.java]: MAJOR Remove this empty statement. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 5 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal " in class " 3 times.](https://github.com/SonarSource/sonar-java/blob/2dabfb90faaff524825ea315dfc363b33ef07e22/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L155) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)
* MAJOR [Rename method "symmetric" to prevent any misunderstanding/clash with field "symmetric" defined on line 90.](https://github.com/SonarSource/sonar-java/blob/2dabfb90faaff524825ea315dfc363b33ef07e22/java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java#L275) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1845)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal " in class " 3 times.](https://github.com/SonarSource/sonar-java/blob/9fe434c50b745a910c49b3fe7d418f9278b4f361/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L155) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)
* MAJOR [Rename method "symmetric" to prevent any misunderstanding/clash with field "symmetric" defined on line 90.](https://github.com/SonarSource/sonar-java/blob/9fe434c50b745a910c49b3fe7d418f9278b4f361/java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java#L275) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1845)

File: java-squid/src/test/files/se/MaxSteps.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    a &= (b() == C);
+    a &= (b() == C);
+    a &= (b() == C);
+    a &= (b() == C);
+    a &= (b() == C);
+
+    if (a) { //BOOM : 2^n -1 states are generated (where n is the number of lines of &= assignements in the above code) -> fail fast by not even enqueuing nodes
+    }
+  }
+}
```
Comment (mpaladin) [inline java-squid/src/test/files/se/MaxSteps.java]: @benzonico EOF new line.

**Comment by mpaladin** [comment]
@benzonico `if(` formatting -> `if (`

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal " in class " 3 times.](https://github.com/SonarSource/sonar-java/blob/5667ae479c137b63d97d7090903c1ec2b538de6e/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L155) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)
* MAJOR [Rename method "symmetric" to prevent any misunderstanding/clash with field "symmetric" defined on line 90.](https://github.com/SonarSource/sonar-java/blob/5667ae479c137b63d97d7090903c1ec2b538de6e/java-squid/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java#L275) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1845)

**Pull request merged**